lib/IRGen/GenStruct.cpp

```diff
@@ -337,6 +339,43 @@ namespace {
                                    const ClangFieldInfo &field) const {
       llvm_unreachable("non-fixed field in Clang type?");
     }
+
+    void initializeWithCopy(IRGenFunction &IGF, Address destAddr,
+                            Address srcAddr, SILType T, bool isOutlined) const {
+      const clang::CXXConstructorDecl *copyConstructor;
+      for (auto member : T.getStructOrBoundGenericStruct()->getMembers()) {
+        if (auto ctor =
+                dyn_cast<clang::CXXConstructorDecl>(member->getClangDecl())) {
+          if (ctor->isCopyConstructor()) {
+            copyConstructor = ctor;
+            break;
+          }
+        }
+      }
+      // TODO: here we should look for the copy asignment operator, otherwise
+      // we'll emit a memcpy when we should call a "real" copy constructor.
+      if (!copyConstructor || copyConstructor->isImplicit() ||
+          copyConstructor->isDefaulted()) {
+        StructTypeInfoBase<ClangRecordTypeInfo, LoadableTypeInfo,
+                           ClangFieldInfo>::initializeWithCopy(IGF, destAddr,
+                                                               srcAddr, T,
+                                                               isOutlined);
+        return;
+      }
+
+      auto globalDecl =
+          clang::GlobalDecl(copyConstructor, clang::Ctor_Complete);
+      auto clangFnAddr =
+          IGF.IGM.getAddrOfClangGlobalDecl(globalDecl, NotForDefinition);
+      auto callee = cast<llvm::Function>(clangFnAddr->stripPointerCasts());
+      auto dest = IGF.coerceValue(destAddr.getAddress(),
+                                  callee->getFunctionType()->getParamType(0),
+                                  IGF.IGM.DataLayout);
+      auto src = IGF.coerceValue(srcAddr.getAddress(),
+                                 callee->getFunctionType()->getParamType(1),
+                                 IGF.IGM.DataLayout);
+      IGF.Builder.CreateCall(callee, {dest, src});
```

Comment:
I created a helper function and generalized `emitConstructorThunkIfNeeded` a bit (to take a `Signature` instead of a `SILFunction` and a `StringRef name` instead of `LinkEntity`).

Reply:
Looking at the tests from #30630. I'm getting:
```
 error: unable to load standard library for target 'x86_64-unknown-windows-msvc'
```
What target builds the windows stdlib?

Reply:
> I created a helper function and generalized `emitConstructorThunkIfNeeded` a bit (to take a `Signature` instead of a `SILFunction` and a `StringRef name` instead of `LinkEntity`). Thanks for pointing this out :)

Hm. I had forgotten that `emitCXXConstructorThunkIfNeeded` takes a `SILFunction`.

I think that the way you're calling the modified function doesn't work though. I think the way you're calling it, `assumedFnType` will always be equal to `ctorFnType`, so it will never create a thunk. What I think you need to do is to create your own assumed function type, then compare against that.

Maybe, at the end of the day, it's easier to do things at the SIL level after all though -- i.e. import the copy constructor as a hidden initializer, emit a call to that, and let the existing machinery take care of creating a thunk if needed.

Reply:
> @martinboehme Looking at the tests from #30630. I'm getting:
> 
> ```
>  error: unable to load standard library for target 'x86_64-unknown-windows-msvc'
> ```
> 
> What target builds the windows stdlib?

Not sure what you mean -- are you getting this error when running existing tests on #30630 (I think they should all be passing on CI though?) or when trying to create new tests in the spirit of the tests in #30630?

It's unfortunately not possible currently to cross-compile stdlib, IIUC. In other words, you can't create a Windows stdlib on Linux, for example. It would be great if this were possible, but for the time being, if you want to do portable IRGen tests (i.e. run Windows IRGen tests on Linux, for example), you can't use stdlib in them. This is pretty restrictive, but you can usually still find a way to test what you want to test -- for an example, see `constructors-irgen.swift` in #30630.

Reply:
> I think that the way you're calling the modified function doesn't work though. I think the way you're calling it, assumedFnType will always be equal to ctorFnType, so it will never create a thunk. What I think you need to do is to create your own assumed function type, then compare against that.

I think you're right.

> Maybe, at the end of the day, it's easier to do things at the SIL level after all though -- i.e. import the copy constructor as a hidden initializer, emit a call to that, and let the existing machinery take care of creating a thunk if needed.

We could also import it as a `__double_underscore` method which should help with overload resolution. Then just use it's function signature and ignore it. 

Alternatively, we could try to build the "correct" function type in IRGen and then use that in the signature argument.